### PR TITLE
fix: Some fixes and improvements to vite-plugin-cypress-esm

### PR DESCRIPTION
**Pull request by M165437**
We are using this plugin with a large react application and ran into issues. These changes address some of them.

### Additional details
- The plugin seemed to have been renamed in the past. I updated some occurrences of the previous name `vite-plugin-mock-esm`. The debug string mentioned in the README now actually works: `DEBUG=cypress:vite-plugin-cypress-esm`.
- The regex capture group for the import target included quotes and semicolon. I updated it to only capture what is between the quotes. For example, in our application, we import a named export "from" (`import { from } from 'myPackage'`). This caused problems. The updated regex now handles this case.
- Additionally, I'm now removing a possible leading dot slash from the import target which further improves picomatch behaviour (see https://github.com/micromatch/picomatch/issues/77).

### How has the user experience changed?
I corrected the picomatch glob matching. It's not trying to match `"./folder/file";` anymore but correctly (bash spec conform) the actual path `folder/file`. This might lead to matching failures for some people who wrote glob patterns for the previous string (that included the quotes and semicolon). Still, I believe this change is necessary as it's the correct way to match globs.

### CODE REVIEW

File: npm/vite-plugin-cypress-esm/src/index.ts [hunk 1]
```diff
@@ -120,21 +118,30 @@ export const CypressEsm = (options?: CypressEsmOptions): Plugin => {
 
     // Ensure import comes at start of line *or* is prefixed by a space so we don't capture things like
     // `Refresh.__hmr_import('')
-    const importRegex = /(?<=^|\s)import (.+?) from (.*)/g
+    const importRegex = /(?<=^|\s)import (.+?) from (['"].*)/g
```
Comment (M165437) [inline npm/vite-plugin-cypress-esm/src/index.ts]: We have an import `import { from } from 'myPackage'` that caused the detected import target to be `} from 'myPackage'`. Now, the capturing group has to start with a quote.

File: npm/vite-plugin-cypress-esm/src/index.ts [hunk 2]
```diff
@@ -120,21 +118,30 @@ export const CypressEsm = (options?: CypressEsmOptions): Plugin => {
...
     return code.replace(
       importRegex,
       (match, importVars: string, importTarget: string) => {
-        if (isImportOnIgnoreList(importTarget)) {
-          debug(`⏭️ Import ${importTarget} matches ignoreImportList, ignoring`)
+        // Strip quotes & semicolons
+        const sanitizedImportTarget = importTarget.replace(/["';]/gi, '').trim()
+
+        if (isValidPath(sanitizedImportTarget)) {
+          debug(`🚫 Invalid import target (${sanitizedImportTarget}), ignoring`)
```
Comment (M165437) [inline npm/vite-plugin-cypress-esm/src/index.ts]: We saw some import statements in comments or console outputs (see example below) which were picked up by this plugin. Now, each import target has to be a valid path.

Example
```
console.warn('(...) You should use `import { unstable_ClassNameGenerator } from '@mui/material/className'` instead (...)');
```

File: npm/vite-plugin-cypress-esm/src/index.ts [hunk 3]
```diff
@@ -71,24 +72,21 @@ export const CypressEsm = (options?: CypressEsmOptions): Plugin => {
...
    * If a given import target (path) is explicitly ignored then bypass our custom mapping on it
    *
-   * @param importTarget
+   * @param sanitizedImportTarget
    * @returns
    */
-  const isImportOnIgnoreList = (importTarget: string) => {
-    return ignoreImportMatcher(importTarget)
+  const isImportOnIgnoreList = (sanitizedImportTarget: string) => {
+    return ignoreImportMatcher(sanitizedImportTarget)
```
Comment (M165437) [inline npm/vite-plugin-cypress-esm/src/index.ts]: The import ignore matcher didn't work correctly. Now it does, after removing the quotes and semicolon from the import target.

File: npm/vite-plugin-cypress-esm/src/index.ts [hunk 4]
```diff
@@ -1,9 +1,10 @@
 import debugFn from 'debug'
 import picomatch from 'picomatch'
+import isValidPath from 'is-valid-path'
 import type { Plugin } from 'vite'
 import fs from 'fs'
 import path from 'path'
-const debug = debugFn('cypress:vite-plugin-mock-esm')
+const debug = debugFn('cypress:vite-plugin-cypress-esm')
```
Comment (M165437) [inline npm/vite-plugin-cypress-esm/src/index.ts:6]: Updated to the debug string already mentioned in the README.

**Comment by jennifer-shehane** [comment]
@M165437 Thanks for the contribution! There are some tests failing from these changes, can you take a look? https://app.circleci.com/pipelines/github/cypress-io/cypress/61103/workflows/bb62c8f0-bf9c-4838-aec4-556582a46fdf/jobs/2540478

File: npm/vite-plugin-cypress-esm/src/index.ts (hunk 2 above)
Comment (M165437) [reply npm/vite-plugin-cypress-esm/src/index.ts]: Unfortunately, this doesn't work. I have given up for now to detect import statements inside strings and comments. We will need to strip out this line independently.

File: npm/vite-plugin-cypress-esm/src/index.ts (hunk 3 above)
Comment (M165437) [reply npm/vite-plugin-cypress-esm/src/index.ts]: I further improved the regex to only capture what's between the quotes.

File: npm/vite-plugin-cypress-esm/src/index.ts (hunk 1 above)
Comment (M165437) [reply npm/vite-plugin-cypress-esm/src/index.ts]: I further improved the regex to only capture what's between the quotes.

File: npm/vite-plugin-cypress-esm/src/index.ts [hunk 5]
```diff
@@ -120,7 +120,7 @@ export const CypressEsm = (options?: CypressEsmOptions): Plugin => {
 
     // Ensure import comes at start of line *or* is prefixed by a space so we don't capture things like
     // `Refresh.__hmr_import('')
-    const importRegex = /(?<=^|\s)import (.+?) from (.*)/g
+    const importRegex = /(?<=^|\s)import (.+?) from ['"](.*?)['"]/g
```
Comment (M165437) [inline npm/vite-plugin-cypress-esm/src/index.ts:123]: Only capture what's between the quotes.

File: npm/vite-plugin-cypress-esm/src/index.ts [hunk 6]
```diff
@@ -139,11 +139,7 @@ export const CypressEsm = (options?: CypressEsmOptions): Plugin => {
           return match
         }
 
-        let replacement = `import * as cypress_${moduleIdentifier}_${++counter} from ${importTarget}`
-
-        if (!replacement.endsWith(';')) {
-          replacement += ';'
-        }
+        let replacement = `import * as cypress_${moduleIdentifier}_${++counter} from '${importTarget}';`
```
Comment (M165437) [inline npm/vite-plugin-cypress-esm/src/index.ts:142]: Add quotes and semicolon as they are not included in the import target anymore.

File: npm/vite-plugin-cypress-esm/src/index.ts [hunk 7]
```diff
@@ -75,7 +75,10 @@ export const CypressEsm = (options?: CypressEsmOptions): Plugin => {
    * @returns
    */
   const isImportOnIgnoreList = (importTarget: string) => {
-    return ignoreImportMatcher(importTarget)
+    // Remove leading dot slash (https://github.com/micromatch/picomatch/issues/77)
+    const importTargetPath = importTarget.replace(/^\.\//, '')
```
Comment (M165437) [inline npm/vite-plugin-cypress-esm/src/index.ts:79]: Remove possible leading dot slash. See https://github.com/micromatch/picomatch/issues/77

**Comment by M165437** [comment]
@jennifer-shehane The tests are passing now. WDYT? :)

**Comment by jennifer-shehane** [comment]
@M165437 We'll have to assign some reviewers, hopefully this sprint.

File: npm/vite-plugin-cypress-esm/src/index.ts [hunk 8]
```diff
@@ -209,6 +209,12 @@ export const CypressEsm = (options?: CypressEsmOptions): Plugin => {
     return code.replace(
       RE,
       (match, importVars: string, importTarget: string) => {
+        if (isImportOnIgnoreList(importTarget)) {
```
Comment (M165437) [inline npm/vite-plugin-cypress-esm/src/index.ts:212]: One last commit to also ignore _dynamic_ imports present on the `ignoreImportList`

**Review by AtofStryker — APPROVED** [review]

**Review by ryanthemanuel — APPROVED** [review]

**Pull request merged**